compiler/rustc_metadata/src/rmeta/table.rs

```diff
@@ -259,44 +260,38 @@ impl<T> FixedSizeEncoding for Option<LazyArray<T>> {
 }
 
 /// Helper for constructing a table's serialization (also see `Table`).
-pub(super) struct TableBuilder<I: Idx, T>
-where
-    Option<T>: FixedSizeEncoding,
-{
-    blocks: IndexVec<I, <Option<T> as FixedSizeEncoding>::ByteArray>,
+pub(super) struct TableBuilder<I: Idx, T: FixedSizeEncoding> {
+    blocks: IndexVec<I, T::ByteArray>,
     _marker: PhantomData<T>,
 }
 
-impl<I: Idx, T> Default for TableBuilder<I, T>
-where
-    Option<T>: FixedSizeEncoding,
-{
+impl<I: Idx, T: FixedSizeEncoding> Default for TableBuilder<I, T> {
     fn default() -> Self {
         TableBuilder { blocks: Default::default(), _marker: PhantomData }
     }
 }
 
-impl<I: Idx, T> TableBuilder<I, T>
+impl<I: Idx, const N: usize, T> TableBuilder<I, Option<T>>
 where
-    Option<T>: FixedSizeEncoding,
+    Option<T>: FixedSizeEncoding<ByteArray = [u8; N]>,
 {
-    pub(crate) fn set<const N: usize>(&mut self, i: I, value: T)
-    where
-        Option<T>: FixedSizeEncoding<ByteArray = [u8; N]>,
-    {
+    pub(crate) fn set(&mut self, i: I, value: T) {
+        self.set_nullable(i, Some(value))
+    }
+}
+
+impl<I: Idx, const N: usize, T: FixedSizeEncoding<ByteArray = [u8; N]>> TableBuilder<I, T> {
+    pub(crate) fn set_nullable(&mut self, i: I, value: T) {
```

Comment:
I think we should have a convenience wrapper around `set_nullable` for `TableBuilder<I, bool>` to just set it to `true`. Invoking `set_nullable(val, false)` is never done and likely a bug.

Reply:
Together with https://github.com/rust-lang/rust/pull/107166#discussion_r1083537832 I'll try to move the true/false conditions to inside the `set` methods.
So it will look approximately like this
```rust
fn set(value) {
    if value.is_default() {
        return;
    }
    // possibly resize the buffer and set the non-default value
}
```
and `false`/`None`/etc arguments will be accepted but result in doing nothing.

The "reset to `None`" semantic is never currently used for metadata tables, but if it will be needed it will require a separate method like `fn clear` in this case.